packages/block-editor/src/hooks/anchor.js

```diff
@@ -40,12 +40,39 @@ export function addAttribute( settings ) {
 	if ( 'type' in ( settings.attributes?.anchor ?? {} ) ) {
 		return settings;
 	}
+
+	// Gracefully handle if settings.attributes is undefined.
 	if ( hasBlockSupport( settings, 'anchor' ) ) {
-		// Gracefully handle if settings.attributes is undefined.
-		settings.attributes = {
-			...settings.attributes,
-			anchor: ANCHOR_SCHEMA,
-		};
+		let saveElement;
+		try {
+			saveElement =
+				typeof settings.save === 'function'
+					? settings.save()
+					: settings.save;
+			if ( saveElement === null || saveElement === undefined ) {
+				// If the save function returns null or undefined, save the anchor
+				// attribute as a block's comment delimiter without specifying
+				// the source.
+				settings.attributes = {
+					...settings.attributes,
+					anchor: {
+						type: ANCHOR_SCHEMA.type,
+					},
+				};
+			} else {
+				settings.attributes = {
+					...settings.attributes,
+					anchor: ANCHOR_SCHEMA,
+				};
+			}
+		} catch ( error ) {
+			// If the save function returns an error, the anchor will be saved
+			// in the markup as an id attribute.
+			settings.attributes = {
+				...settings.attributes,
+				anchor: ANCHOR_SCHEMA,
+			};
+		}
```

Comment:
This is probably the most complicated part, but depending on the result of the value of the `save` property, I register the `anchor` attribute with the appropriate definition.

Reply:
It does feel a bit complicated.

I'd probably be more in favor of something declarative on the blocks that need it rather than magically detecting how it should work.

i.e.
```js
supports: {
    anchor: { attributeSource: 'html'|'delimiter' } // defaults to 'html' when `anchor: true`
}
```

Another possible option is to migrate static blocks to use the block delimiter attribute instead of the `id` html attribute, so that it works the same on client and server.